content/en/blog/_posts/2022-TBD-contextual-logging/index.md

```diff
@@ -0,0 +1,248 @@
+with quite a bit
+of text. internal:0}
+I1112 14:06:35.783549  328441 structured_logging.go:52] "using InfoS with\nthe message across multiple lines" int=1 stringData="long: Multiple\nlines\nwith quite a bit\nof text." str="another value"
+```
+
+Now `<` and `>` markers and indention are used to ensure that splitting at a
+klog header at the start of a line is reliable and the output remains readable
+for humans:
+
+```
+I1126 10:31:50.378204  121736 structured_logging.go:59] "using InfoS" longData=<
+	{Name:long Data:Multiple
+	lines
+	with quite a bit
+	of text. internal:0}
+ >
+I1126 10:31:50.378228  121736 structured_logging.go:60] "using InfoS with\nthe message across multiple lines" int=1 stringData=<
+	long: Multiple
+	lines
+	with quite a bit
+	of text.
+ > str="another value"
+```
+
+Note that the log message itself is printed with quoting. It is meant to be a
+fixed string that identifies a log entry, so newlines should be avoided there.
+
+Before Kubernetes 1.24, some log calls in kube-scheduler still used `klog.Info`
+for multi-line strings to avoid the unreadable output. Now kube-scheduler has
+been converted completely to structured logging.
+
+## Contextual Logging
+
+[Contextual
+logging](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/3077-contextual-logging/README.md)
+is based on the [go-logr
+API](https://github.com/go-logr/logr#a-minimal-logging-api-for-go). The key
+idea is that libraries are passed a logger instance by their caller and use
+that for logging instead of accessing a global logger. The binary decides about
+the logging implementation, not the libraries. The go-logr API is designed
+around structured logging and supports attaching additional information to a
+logger.
+
+This enables additional use cases:
+
+- The caller can attach a prefix
+  ([`WithName`](https://pkg.go.dev/github.com/go-logr/logr#Logger.WithName))
+  and additional values
+  ([`WithValues`](https://pkg.go.dev/github.com/go-logr/logr#Logger.WithValues))
+  to the logger that it passes into a function. This additional information is
+  then included in all log entries, without having to modify the code that
+  generates the log entries. This is useful in highly parallel applications
+  where it can become hard to identify all log entries for a certain operation
+  because the output from different operations gets interleaved.
+
+- When running unit tests, log output can be associated with the current test.
+  Then when a test fails, only the log output of the failed test gets shown
+  by `go test`. That output can also be more verbose by default because it
+  will not get shown for successful tests. Tests can be run in parallel
+  without interleaving their output.
+
+Code has to be modified to take advantage of contextual logging. The
+traditional klog calls for logging continue to work unchanged but then don't
+support contextual logging because output still goes through the global klog
+logger. The [migration
```

Comment:
The migration guide doesn't seem to have anything specific about contextual logging. I couldn't locate any specific text dealt with deriving logger from Context or using APIs like FromName etc.

Reply:
I should have mentioned that https://github.com/kubernetes/community/pull/6560 needs to be merged first.

Reply:
/cc @serathius @logicalhan